flink-table/flink-table-planner/src/main/java/org/apache/flink/table/catalog/DatabaseCalciteSchema.java

```diff
@@ -77,6 +84,23 @@ public Table getTable(String tableName) {
 						!connectorTable.isBatch(),
 						FlinkStatistic.of(tableSource.getTableStats().orElse(null))))
 					.orElseThrow(() -> new TableException("Cannot query a sink only table."));
+			} else if (table instanceof CatalogTable) {
+				Optional<TableFactory> tableFactory = catalog.getTableFactory();
+				TableSource<Row> tableSource = tableFactory.map(tf -> ((TableSourceFactory) tf).createTableSource((CatalogTable) table))
```

Comment:
the `tf` in `.map(tf -> xxx)` is an Optional object and cannot be casted to `TableSourceFactory`, right? 

BTW, what happens if the `tableFactory` is not present? Also, maybe I missed it - I only find table source, what about table sink?

Reply:
map() is applied to an Optional object, so the key (tf) is a TableFactory instance.
As you can see from other types of tables, calcite seems only need TableSource.
If tableFactory isn't present, then orElse() clause kicks in. The test verifies that.

Reply:
nice, I didn't know map() can be applied to Optional objects